spotrem: add volume knob gesture

thyttan opened the pull request.

bobrippling commented:
Very cool! Works well when I tested it. Do you think it's worth mentioning that the clockwise/anticlockwise gesture detection is stopped after a brief period of no-interaction?

Not for this PR, but if you wanted ideas on going further with this feature, it'd be cool to pull it out to a library and let other apps opt into the clockwise gesture, and during the clockwise gesture, display a circle on screen that follows the user's gesture. But obviously those are much bigger changes!

thyttan commented:
Thanks :)

> Do you think it's worth mentioning that the clockwise/anticlockwise gesture detection is stopped after a brief period of no-interaction?

Yes! Will add that.

> [...] it'd be cool to pull it out to a library and let other apps opt into the clockwise gesture, and during the clockwise gesture, display a circle on screen that follows the user's gesture. But obviously those are much bigger changes!

Yes - that's where I started with this. But I switched to this focused implementation first to dial in the UX (pun intended).

thyttan commented:
@bobrippling this is maybe done now! Please comment any feedback you may have :)

thyttan commented on modules/Dial_Display.js:
```diff
@@ -0,0 +1,59 @@
+      if (isFill) g.fillCircle(x, y, rad);
+    }
+    if (this.isFirstDraw) {
+      g.setColor(0,0,0).fillCircle(CENTER.x, CENTER.y, 25);
```
Maybe adding a additional `displayRect` setting to the options object that if is not set defaults to what the `dialRect` options is.

bobrippling replied:
Yeah, I think this is ok for now but we could have a follow-up PR which infers a radius from the rect? Or like you say, adds a setting

bobrippling commented on modules/Dial_Display.js:
```diff
@@ -0,0 +1,59 @@
+      let x = CENTER.x+27*Math.sin(value*(2*Math.PI/options.stepsPerWholeTurn));
+      let y = CENTER.y-27*Math.cos(value*(2*Math.PI/options.stepsPerWholeTurn));
+      g.setColor(R,G,B)
+      if (!isFill) g.drawCircle(x, y, rad);
+      if (isFill) g.fillCircle(x, y, rad);
```
Perhaps:

```suggestion
      if (!isFill) g.drawCircle(x, y, rad);
      else g.fillCircle(x, y, rad);
```

bobrippling replied:
Ah gotya - I'll let you decide on `drawCircle`'s fate just in case you do find a use

bobrippling commented on modules/Dial_Display.md:
````diff
@@ -0,0 +1,65 @@
+
+```JS
+var DialDisplay = require("Dial_Display");
+var dialDisplay = new DialDisplay(options);
````
Same again, but here I think we'll want to change to make it more class-like, or alter how `this` is used within the `dialDisplay` function. At the moment it'll be writing to the global object.

For example, in the IDE:
```javascript
> cb
=function (step) { ... }
>cb(1)
=undefined
>cb(1)
=undefined

// the dial renders successfully
// `this` variables are written to the global:

>isFirstDraw
=false
>value
=14
```

Happy to go through a few suggestions for it, if you like

bobrippling replied:
Of course, I'll pop something together

bobrippling replied:
[Here you go](https://github.com/espruino/BangleApps/pull/3847/commits/91a6b9d78638c7ea3f7737c877a18aa7c144d4b6?diff=split&w=1) - I've summarised the usage in the updated readme (inline below), let me know what you think. Happy to revert if you'd prefer a different approach :)


https://github.com/espruino/BangleApps/blob/91a6b9d78638c7ea3f7737c877a18aa7c144d4b6/modules/Dial_Display.md#L11-L20

bobrippling replied:
Yes very true, `reset` is a little ambiguous! (feel free to change it, delete any of the methods or anything of course)

But yes - I found exactly what you said, `set()` makes it forget the previous value, so I altered the code slightly to deal with that too

thyttan replied:
> so I altered the code slightly to deal with that too

I don't think you pushed this? 🙂

bobrippling replied:
Ah, sorry I should've said - it was all in the same commit, the code keeping `prevValue` in sync is here :)

https://github.com/espruino/BangleApps/blob/91a6b9d78638c7ea3f7737c877a18aa7c144d4b6/modules/Dial_Display.js#L24-L32

bobrippling reviewed (commented):
> @bobrippling this is maybe done now! Please comment any feedback you may have :)
> 
> It now adds two new modules for dial functionality, one for touch interaction and one for graphics. This is used in `spotrem` to provide the volume knob.

Great, nice separation of functionality! I spotted a few things, let me know if I've been unclear or if you've any questions :)

thyttan commented:
@bobrippling would you mind having a look again? :) 

I fixed an edge case where `Dial_Display` could lose track of the previous position and not clear the indicator there. Also renamed the `reset` method to `queueRedraw`.

bobrippling commented:
Of course - nice fix! That code's a lot more robust feeling too, LGTM :)
